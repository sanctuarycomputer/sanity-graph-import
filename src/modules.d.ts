declare module 'prompt-confirm'
declare module '@sanity/import/*'
