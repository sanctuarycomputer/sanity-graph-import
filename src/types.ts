import {
  SanityAssetDocument as SanityAssetDocumentType,
  SanityDocument as SanityDocumentType
} from '@sanity/client';

export type Maybe<T> = T | null | undefined | void;

export type SanityAssetDocument = SanityAssetDocumentType;

export interface SanityObject {
  // _type: string;
  [key: string]: SanityFieldValue;
}

export type SanityFieldValue =
  | SanityFieldValue[]
  | SanityReference
  | SanityObject
  | SanityAssetObject
  | boolean
  | string
  | number
  | void;

export type SanityDocument = Pick<SanityDocumentType, '_id' | '_type'> & {
  [key: string]: SanityFieldValue;
};

export interface MigratedDocument<
  DocType extends SanityDocument = SanityDocument
> {
  source: DocType;
  destination: DocType;
}

export interface UnMigratedDocument<
  DocType extends SanityDocument = SanityDocument
> {
  source: DocType;
  destination?: DocType;
}

export type MigratedAsset = MigratedDocument<SanityAssetDocument>;
export type UnMigratedAsset = UnMigratedDocument<SanityAssetDocument>;

export interface SanityReference extends SanityObject {
  _type: 'reference';
  _ref: string;
}

export interface SanityAssetObject extends SanityObject {
  _type: 'image' | 'file';
  asset: SanityReference;
}
