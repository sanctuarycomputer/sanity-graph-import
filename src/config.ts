export const DEFAULT_BATCH_SIZE = 35
