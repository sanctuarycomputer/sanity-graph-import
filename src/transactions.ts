import {
  SanityClient,
  SanityAssetDocument,
  Transaction,
  MultipleMutationResult,
} from '@sanity/client'
import getHashedBufferForUri from '@sanity/import/src/util/getHashedBufferForUri'
import { MigratedAsset, UnMigratedAsset, SanityDocument } from './types'
import {
  definitely,
  logWrite,
  logDelete,
  partition,
  getImageHash,
  getAssetType,
  isMigratedDocument,
  getUploadedFilename,
  chunk,
  createRemapReferences,
  queue,
} from './utils'

export const deleteAll = async (
  client: SanityClient
): Promise<MultipleMutationResult> => {
  const { dataset, projectId } = client.config()
  logDelete(`Clearing ${projectId}/${dataset} dataset...`)
  const allDocuments = await client.fetch<SanityDocument[]>('*[]')
  const ids = allDocuments
    .map((doc) => doc._id)
    .filter(
      (id) =>
        /* Reserved and internal documents start with "_." Omit these. */
        !/^_\./.test(id)
    )
  const initialTrx = client.transaction()
  const finalTrx = ids.reduce<Transaction>(
    (previousTrx, id) => previousTrx.delete(id),
    initialTrx
  )
  const result = await finalTrx.commit()
  logDelete(`Removed ${ids.length} documents from ${projectId}/${dataset}`)
  return result
}

export const uploadAsset = async (
  client: SanityClient,
  originalAsset: SanityAssetDocument
): Promise<MigratedAsset> => {
  const url = originalAsset.url
  const assetType = getAssetType(originalAsset)
  const { buffer } = await getHashedBufferForUri(url)

  const options = {
    label: getImageHash(originalAsset),
    filename: getUploadedFilename(originalAsset),
    source: {
      id: originalAsset._id,
      source: 'migration',
      url: originalAsset.url,
    },
  }
  const uploadResult = await client.assets.upload(
    // @ts-ignore not sure how to enforce the string type for the overloads
    assetType,
    buffer,
    options
  )
  return {
    source: originalAsset,
    destination: uploadResult,
  }
}

export const uploadAssets = async (
  client: SanityClient,
  sourceAssets: SanityAssetDocument[]
): Promise<MigratedAsset[]> => {
  const originalHashes = definitely(sourceAssets.map(getImageHash))

  /* Find assets that already exist */
  const existingAssets = await client.fetch<SanityAssetDocument[]>(
    `*[label in $originalHashes]`,
    { originalHashes }
  )

  const assetPairs = sourceAssets.map<MigratedAsset | UnMigratedAsset>(
    (source) => ({
      source,
      destination: existingAssets.find(
        (existingAsset) => existingAsset.label === getImageHash(source)
      ),
    })
  )

  const [migrated, unmigrated] = partition<MigratedAsset, UnMigratedAsset>(
    assetPairs,
    isMigratedDocument
  )

  if (unmigrated.length === 0) {
    logWrite('Found no new assets to upload')
    return migrated
  }

  logWrite(`Found ${unmigrated.length} new assets to upload`)

  const [newAssets, errors] = await queue(
    unmigrated.map(({ source }) => () => uploadAsset(client, source))
  )

  if (errors.length) {
    errors.forEach((error) => {
      console.log(error)
    })
  }

  return [...migrated, ...newAssets]
}

interface InsertDocumentOptions {
  batchSize: number
}

export const insertDocuments = async (
  client: SanityClient,
  sourceDocuments: SanityDocument[],
  sourceAssets: SanityAssetDocument[],
  { batchSize }: InsertDocumentOptions
) => {
  const dataset = client.config().dataset
  const uploadedAssets = await uploadAssets(client, sourceAssets)

  const remapRefs = createRemapReferences(sourceDocuments, uploadedAssets)
  const updatedDocuments = sourceDocuments.map((doc) => remapRefs(doc))

  const insertBatch = async (
    batch: SanityDocument[]
  ): Promise<MultipleMutationResult> => {
    const transaction = batch.reduce<Transaction>(
      (prevTrx, document) => prevTrx.createOrReplace(document),
      client.transaction()
    )

    return transaction.commit()
  }

  const documentBatches = chunk(updatedDocuments, batchSize)

  logWrite(
    `Inserting ${sourceDocuments.length} documents into dataset "${dataset}"`
  )

  const [results, errors] = await queue(
    documentBatches.map((batch) => () => insertBatch(batch))
  )

  if (errors.length) {
    errors.forEach((error) => {
      console.log(error)
    })
  }

  logWrite(
    `Inserted ${updatedDocuments.length} documents in ${results.length} batches`
  )
  const updatedStrong = sourceDocuments.map((doc) => remapRefs(doc, false))
  const strongBatches = chunk(updatedStrong, batchSize)

  logWrite('Strengthening references..')

  const [strongResults, strongResultErrors] = await queue(
    strongBatches.map((batch) => () => insertBatch(batch))
  )

  if (strongResultErrors.length) {
    strongResultErrors.forEach((error) => {
      console.log(error)
    })
  }

  return strongResults
}
