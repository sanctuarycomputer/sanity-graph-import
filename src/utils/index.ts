export * from './catchable'
export * from './misc'
export * from './parsing'
export * from './logging'
export * from './queue'
export * from './retryCaughtErrors'
