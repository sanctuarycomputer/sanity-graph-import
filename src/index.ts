import type { SanityClient } from '@sanity/client'
import PromptConfirm from 'prompt-confirm'
import invariant from 'tiny-invariant'
import { log, flat, unique, findReferencedIds, logFetch, queue } from './utils'
import { deleteAll, insertDocuments } from './transactions'
import { SanityDocument, SanityAssetDocument } from './types'
import { DEFAULT_BATCH_SIZE } from './config'

type QueryParams = Record<string, any>

interface QueryArgs {
  query: string
  params?: QueryParams
}

interface SourceConfig {
  initialQueries: QueryArgs[]
  client: SanityClient
}

interface DestinationConfig {
  /* The destination client must have a write token! */
  client: SanityClient
  /**
   * The number of documents to include in a batch.
   *
   * default: 35
   *
   * If you are getting 'content-length' errors during migration,
   * set this to a lower number.
   */
  batchSize?: number
}

interface ImportConfig {
  source: SourceConfig
  destination: DestinationConfig
}

const getConfigString = (client: SanityClient): string => {
  const { projectId, dataset } = client.config()
  if (!projectId) invariant('Your client must include a projectId')
  if (!dataset) invariant('Your client must include a dataset')
  return [projectId, dataset].join(':')
}

export const migrate = async ({
  source,
  destination,
}: ImportConfig): Promise<void> => {
  const sourceClient = source.client
  const destinationClient = destination.client

  /**
   * Validate config
   */
  const sourceConfigString = getConfigString(sourceClient)
  const destConfigString = getConfigString(destinationClient)

  invariant(
    !(sourceConfigString === destConfigString),
    `Both clients have the same configuration: ${sourceConfigString}`
  )

  invariant(
    Boolean(destinationClient.config().token),
    'The destination client must have a write token'
  )

  invariant(
    !Boolean(sourceClient.config().token),
    'The source client must not have a token'
  )

  invariant(
    source.initialQueries.length !== 0,
    'You must include at least one initial query'
  )

  /**
   * Fetch initial documents
   */

  log(
    `Migrating from ${sourceClient.config().projectId}/${
      sourceClient.config().dataset
    } to ${destinationClient.config().projectId}/${
      destinationClient.config().dataset
    } `
  )
  const initialDocuments = flat(
    await queue<SanityDocument[]>(
      source.initialQueries.map(({ query, params }) => () =>
        sourceClient.fetch(query, params || {})
      )
    )
  )

  const docIds = initialDocuments.map(({ _id }) => _id)
  const referencedIds = flat(initialDocuments.map(findReferencedIds))
  const allIds = unique(docIds.concat(referencedIds))

  logFetch(`Found ${initialDocuments.length} initial documents`)
  logFetch(`   ${referencedIds.length} referenced documents`)

  const sourceDocuments = await sourceClient.fetch<SanityDocument[]>(
    `*[_id in $allIds && _type != 'sanity.imageAsset' && _type != 'sanity.fileAsset']`,
    { allIds }
  )
  logFetch(`Fetched all referenced documents`)

  const assetIds = flat(sourceDocuments.map(findReferencedIds)).filter(
    (id) => id.startsWith('image-') || id.startsWith('file-')
  )
  const sourceAssets = await sourceClient.fetch<SanityAssetDocument[]>(
    `*[_id in $assetIds]`,
    { assetIds }
  )
  logFetch(`Fetched ${sourceAssets.length} source assets`)

  const confirmDelete = new PromptConfirm(
    'Do you want to remove all data from the destination dataset?'
  )
  const confirmed = await confirmDelete.run()
  if (confirmed) {
    await deleteAll(destinationClient)
  }

  const batchSize = destination.batchSize ?? DEFAULT_BATCH_SIZE

  await insertDocuments(destinationClient, sourceDocuments, sourceAssets, {
    batchSize,
  })

  log('Success! ðŸŽ‰')
}
