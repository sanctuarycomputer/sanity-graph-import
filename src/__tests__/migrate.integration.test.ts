import CreateClient from '@sanity/client'
import { migrate } from '../../src'
import dotenv from 'dotenv'
dotenv.config()

const sourceClient = CreateClient({
  projectId: process.env.TEST_STUDIO_SOURCE_PROJECTID,
  dataset: process.env.TEST_STUDIO_SOURCE_DATASET,
})

const destinationClient = CreateClient({
  projectId: process.env.TEST_STUDIO_TARGET_PROJECTID,
  dataset: process.env.TEST_STUDIO_TARGET_DATASET,
  token: process.env.TEST_STUDIO_TARGET_TOKEN,
})

const initialQuer

const main = async () => {}

main()
