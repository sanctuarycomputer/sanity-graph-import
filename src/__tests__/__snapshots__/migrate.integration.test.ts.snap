// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Full migration should print the same output given the same HTTP requests 1`] = `
Array [
  "Migrating from 98nzfixt/production to 98nzfixt/staging ",
]
`;

exports[`Full migration should print the same output given the same HTTP requests 2`] = `
Array [
  "Fetching 1 initial queries..",
]
`;

exports[`Full migration should print the same output given the same HTTP requests 3`] = `
Array [
  "Found 5 initial documents",
]
`;

exports[`Full migration should print the same output given the same HTTP requests 4`] = `
Array [
  "    movie: 5",
]
`;

exports[`Full migration should print the same output given the same HTTP requests 5`] = `
Array [
  "    + 78 referenced documents",
]
`;

exports[`Full migration should print the same output given the same HTTP requests 6`] = `
Array [
  "Fetching referenced documents in 16 batches..",
]
`;

exports[`Full migration should print the same output given the same HTTP requests 7`] = `
Array [
  "Fetched all referenced documents",
]
`;

exports[`Full migration should print the same output given the same HTTP requests 8`] = `
Array [
  "      + 63 source assets",
]
`;

exports[`Full migration should print the same output given the same HTTP requests 9`] = `
Array [
  "Clearing 98nzfixt/staging dataset...",
]
`;

exports[`Full migration should print the same output given the same HTTP requests 10`] = `
Array [
  "Removed 51 documents from 98nzfixt/staging",
]
`;

exports[`Full migration should print the same output given the same HTTP requests 11`] = `
Array [
  "Found 63 new assets to upload",
]
`;

exports[`Full migration should print the same output given the same HTTP requests 12`] = `
Array [
  "Inserting 74 documents into dataset \\"staging\\"",
]
`;

exports[`Full migration should print the same output given the same HTTP requests 13`] = `
Array [
  "Inserted 74 documents in 3 batches",
]
`;

exports[`Full migration should print the same output given the same HTTP requests 14`] = `
Array [
  "Strengthening references..",
]
`;

exports[`Full migration should print the same output given the same HTTP requests 15`] = `
Array [
  "Success! ðŸŽ‰",
]
`;
